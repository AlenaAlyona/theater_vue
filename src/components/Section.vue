<template>
  <div class="sectionContainer">
    <div class="sectionTitle">{{ section.name }}</div>
    <div class="sectionLayout">
      <table>
        <thead>
          <tr>
            <th class="thRow">row</th>
            <th colspan="5">seats</th>
          </tr>
        </thead>
        <tbody>
          <tr v-for="row in dataToDisplay.rows" :key="row.row">
            <th scope="row" class="sectionRow">{{ row.row }}</th>
            <div class="seatsRow">
              <td
                v-bind:style="booked(seat)"
                v-for="seat in row.seats"
                :key="seat.seat"
                class="sectionSeat"
              >
                {{ seat.group }}
              </td>
            </div>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</template>

<script>
export default {
  name: "Section",
  props: ["section"],

  computed: {
    dataToDisplay: function() {
      const result = [...this.section.rows];
      this.section.groups.forEach((group) => {
        group.seats.forEach((seatInGroup) => {
          const row = result[seatInGroup.row - 1];
          row.seats.forEach((seat) => {
            if (seat.seat === seatInGroup.seat) {
              seat.group = group.id;
              seat.groupColor = group.color;
            }
          });
        });
      });

      return { ...this.section, rows: result };
    },
  },
  methods: {
    booked: function(seat) {
      if (Object.prototype.hasOwnProperty.call(seat, "group")) {
        return {
          background: `${seat.groupColor}`,
        };
      } else if (seat.rank === "rank1") {
        return {
          background: "#e1dde0",
        };
      } else if (seat.rank === "rank2") {
        return {
          background: "#9C9C9C",
        };
      }
    },
  },
};
</script>

<style scoped>
.sectionContainer {
  width: 40%;
  margin: auto;
  margin-bottom: 2rem;
}

.sectionTitle {
  text-transform: capitalize;
  font-weight: 400;
  font-size: 1.5rem;
  margin-bottom: 0.75rem;
}

.sectionLayout {
  border: 1px solid #fff;
  justify-content: center;
  width: 100%;
}

table {
  width: 100%;
  padding: 1rem 0;
}

.thRow {
  width: 4rem;
  padding-left: 0;
}

.sectionRow {
  width: 4rem;
  height: 2rem;
  margin: auto;
}

.seatsRow {
  width: 100%;
  height: 100%;
  display: inline-flex;
  justify-content: center;
  text-align: center;
  flex-shrink: 1;
  flex-wrap: wrap;
}

.sectionSeat {
  width: 1.5rem;
  height: 1.5rem;
  text-align: center;
  color: #202530;
  display: flex;
  margin: 0.7rem;
}
</style>
