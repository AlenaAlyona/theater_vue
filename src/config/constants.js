/* eslint-disable */

export const layout = {
  ranks: ["rank1", "rank2", "rank3"],
  sections: [
    {
      name: "main hall",
      rows: [
        {
          row: "1",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "4", rank: "rank1" },
            { seat: "2", rank: "rank1" },
          ],
        },
        {
          row: "2",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "5", rank: "rank1" },
            { seat: "6", rank: "rank1" },
            { seat: "4", rank: "rank1" },
            { seat: "2", rank: "rank1" },
          ],
        },
        {
          row: "3",
          seats: [
            { seat: "1", rank: "rank2" },
            { seat: "3", rank: "rank2" },
            { seat: "5", rank: "rank2" },
            { seat: "6", rank: "rank2" },
            { seat: "4", rank: "rank2" },
            { seat: "2", rank: "rank2" },
          ],
        },
      ],
    },
    {
      name: "balcony",
      rows: [
        {
          row: "1",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "4", rank: "rank1" },
            { seat: "2", rank: "rank1" },
          ],
        },
        {
          row: "2",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "5", rank: "rank1" },
            { seat: "6", rank: "rank1" },
            { seat: "4", rank: "rank1" },
            { seat: "2", rank: "rank1" },
          ],
        },
        {
          row: "3",
          seats: [
            { seat: "1", rank: "rank2" },
            { seat: "3", rank: "rank2" },
            { seat: "5", rank: "rank2" },
            { seat: "6", rank: "rank2" },
            { seat: "4", rank: "rank2" },
            { seat: "2", rank: "rank2" },
          ],
        },
      ],
    },
  ],
};

export const groups = [
  {
    id: "+31611111111",
    seats: [
      { section: "main hall", row: "1", seat: "4" },
      { section: "main hall", row: "1", seat: "2" },
      { section: "main hall", row: "2", seat: "2" },
      { section: "main hall", row: "2", seat: "4" },
    ],
  },

  {
    id: "+31622222222",
    seats: [
      { section: "main hall", row: "2", seat: "6" },
      { section: "main hall", row: "2", seat: "5" },
    ],
  },

  {
    id: "+31633333333",
    seats: [
      { section: "balcony", row: "3", seat: "6" },
      { section: "balcony", row: "3", seat: "5" },
      { section: "balcony", row: "3", seat: "" },
    ],
  },
];
