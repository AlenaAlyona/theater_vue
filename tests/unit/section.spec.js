import Section from "@/components/Section";
import { testProps } from "../../src/config/constants";
import { mount } from "@vue/test-utils";

describe("Section.vue", () => {
  test("mount a Section component", () => {
    const wrapper = mount(Section, {
      propsData: {
        section: testProps,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("computed", () => {
  test("render seats", () => {
    const { vm } = mount(Section, {
      propsData: {
        section: testProps,
      },
    });

    vm.dataToDisplay.rows[0].seats.forEach((seat) => {
      expect(seat).toHaveProperty("group");
      expect(seat).toHaveProperty("groupColor");
    });

    vm.dataToDisplay.rows[1].seats.forEach((seat) => {
      expect(seat).not.toHaveProperty("group");
      expect(seat).not.toHaveProperty("groupColor");
    });
  });
});

describe("method", () => {
  const { vm } = mount(Section, {
    propsData: {
      section: testProps,
    },
  });

  test("booked", () => {
    expect(
      vm.booked({
        seat: "1",
        rank: "rank1",
        group: "gr2",
        groupColor: "hsla(277,70%,40%,1)",
      })
    ).toStrictEqual({ background: "hsla(277,70%,40%,1)" });

    expect(vm.booked({ seat: "1", rank: "rank1" })).toStrictEqual({
      background: "#e1dde0",
    });

    expect(vm.booked({ seat: "5", rank: "rank2" })).toStrictEqual({
      background: "#9C9C9C",
    });
  });
});
