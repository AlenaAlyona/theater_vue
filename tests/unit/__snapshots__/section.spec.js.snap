// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section.vue mount a Section component 1`] = `
<div class="sectionContainer">
  <div class="sectionTitle">main hall</div>
  <div class="sectionLayout">
    <table>
      <thead>
        <tr>
          <th class="thRow">row</th>
          <th colspan="5">seats</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th scope="row" class="sectionRow">1</th>
          <div class="seatsRow">
            <td style="background: rgb(31, 31, 31);" class="sectionSeat">gr2</td>
            <td style="background: rgb(31, 31, 31);" class="sectionSeat">gr2</td>
            <td style="background: rgb(31, 31, 31);" class="sectionSeat">gr1</td>
            <td style="background: rgb(31, 31, 31);" class="sectionSeat">gr1</td>
          </div>
        </tr>
        <tr>
          <th scope="row" class="sectionRow">2</th>
          <div class="seatsRow">
            <td style="background: rgb(225, 221, 224);" class="sectionSeat"></td>
            <td style="background: rgb(225, 221, 224);" class="sectionSeat"></td>
            <td style="background: rgb(156, 156, 156);" class="sectionSeat"></td>
            <td style="background: rgb(156, 156, 156);" class="sectionSeat"></td>
            <td style="background: rgb(156, 156, 156);" class="sectionSeat"></td>
            <td style="background: rgb(156, 156, 156);" class="sectionSeat"></td>
          </div>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
